### ISSUE TITLE
Wrap entire application (including webpack runtime) w/ error catcher so can recover gracefully from service worker problems

### ISSUE THREAD

**Issue by KyleAMathews** [gatsbyjs/gatsby#11240]
As a failsafe against the webpack runtime not being able to load some JS due to e.g. a js file being deleted from the sw cache, we should wrap the webpack runtime with a try/catch to force a refresh in these sorts of cases.

**Comment by sidharthachatterjee** [comment]
Could potentially address https://github.com/gatsbyjs/gatsby/issues/11194

**Comment by davidbailey00** [comment]
I've had a look into this - since the issue was opened, I'm no longer sure if this will be useful.

The rationale for opening this issue was that sometimes, upon upgrade of a SW, certain Webpack chunks would be unavailable, causing the error "Loading chunk <no> failed" originating from the Webpack runtime.

I've tried reproducing this error in order to make sure that the issue is still relevant, but haven't been able to do so:
- Removing either the app bundle or Webpack runtime bundle simply prevents Gatsby's runtime kicking in entirely - this should be nothing to worry about, especially with PR #13502, which checks for the presence of the correct app bundle before serving a page in app shell mode
- Removing/blocking one or more individual chunks, e.g. https://www.gatsbyjs.org/47-8b65c83151e514b99b9e.js, doesn't cause a Webpack error - this is handled by `ensure-resources.js` which says "Error: Missing resources for /"

So in other words, I fail to see a scenario in which the Webpack runtime could successfully load but an individual chunk fails in a way that we can't catch. As a result, I'm not sure if it would be worthwhile creating a wrapper around this. Thoughts?

**Comment by m-allanson** [comment]
Thanks @David - let's close this for now.

**Comment by ichouu** [comment]
hi,

i was wondering 
"we should wrap the webpack runtime with a try/catch to force a refresh in these sorts of cases."
which file in gatsbys .cache folder does this refer too? 

I am currently having some issues with webpack runtime 404s.

and think this might be a solution? or if anyone knows what else could be going on. 

![Screen Shot 2019-06-21 at 1 44 30 PM](https://user-images.githubusercontent.com/15332786/59953629-dd63e600-9435-11e9-9930-de4ebe58405f.png)

**Comment by mduecker** [comment]
I am still experiencing blank loads (website note being displayed) of my Gatsby app after pushing new updates.
This is not easy to reproduce, but it happens. And it's more likely to occur when the page hasn't been visited for a few days and then a new version is deployed.

Here's a screenshot:
![image](https://user-images.githubusercontent.com/10739903/62196461-5da73080-b37e-11e9-8991-2abc78d4d6a7.png)

In the docs I found this:
[](https://www.gatsbyjs.org/docs/resource-handling-and-service-workers/)
> Blank pages can theoretically occur if we serve the app shell when resources are unavailable - however, this should never occur since we only serve the app shell with whitelisted paths (i.e. ones whose resources have been cached entirely). There may be some edge cases when this can occur, e.g. when the Webpack runtime from the old site attempts to load a chunk which is unavailable on the updated site - we are currently investigating ways to prevent this, and make using service workers with Gatsby even more robust.

Would be good to get a clarifying update on this. How much of an edge case is this? Because I've noticed it multiple times already.

**Comment by tinder-kyleboss** [comment]
Hey everyone 👋 

I would like to add that avid users of our Gatsby sites that are PWAs run into this issue as well. Most of the time it involves the file `/webpack-runtime-[chunk]`.

"Failed to load resource: the server /webpack-runtime-[chunk].js:1 responded with a status of 404 ()"

![image](https://user-images.githubusercontent.com/42850683/70164317-cdebf880-1675-11ea-8176-32dad07c9414.png)

Has there been any success on this front yet?

**Comment by tinder-kyleboss** [comment]
I wanted to report back that the reason I was getting the following error
> "Failed to load resource: the server /webpack-runtime-[chunk].js:1 responded with a status of 404 ()"
was not due to the service worker. It is caused when a user visits the website when, after a build, the new index.html has been uploaded to our server but the new webpack-runtime-CHUNK.js was still in the upload queue.

This may or may not be related to the original issue, but if you find yourself getting this error sporadically, ensure that you are uploading your HTML files to the server only after all the dependencies (page-data/**, webpack-runtime, etc) are uploaded.
